compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3566,11 +3566,12 @@ boolean checkLiteralAssignabilityBasedOnType(BLangLiteral baseLiteral, BLangLite
             case TypeTags.DECIMAL:
                 BigDecimal baseDecimalVal = NumericLiteralSupport.parseBigDecimal(baseValue);
                 BigDecimal candidateDecimalVal;
-                if (candidateTypeTag == TypeTags.INT && !candidateLiteral.isConstant) {
+                if (candidateTypeTag == TypeTags.INT && !candidateLiteral.isConstant &&
+                        !candidateLiteral.isFiniteContext) {
```

Comment:
Thinking out loud, `isFiniteContext` was introduced more for type checking literal members of a finite type. I think the fix for this issue should be in `isAssignableToFiniteType` instead. Since the types for the value space members are already set, can't we check if the types are compatible first and call `checkLiteralAssignabilityBasedOnType` only if compatible?

Reply:
Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/35151/commits/db82396485ab3a71974d8792d96880343a4efe71